boxtree/tree_build.py

```diff
@@ -46,6 +59,269 @@ class MaxLevelsExceeded(RuntimeError):
+    n_stale_boxes = np.sum(tob.box_levels == np.inf)
+    newn = tob.nboxes - n_stale_boxes
+    from dataclasses import replace
+    return replace(tob, box_parent_ids=tob.box_parent_ids[:newn],
+                   box_child_ids=tob.box_child_ids[:, :newn],
+                   box_levels=tob.box_levels[:newn],
+                   box_centers=tob.box_centers[:, :newn])
+
+
+def refined_and_coarsened(tob: TreeOfBoxes,
+                          refine_flags: Optional[npt.NDArray] = None,
+                          coarsen_flags: Optional[npt.NDArray] = None,
+                          error_on_ignored_flags: bool = True
+                          ) -> TreeOfBoxes:
+    """Make a refined/coarsened copy. When children of the same parent box
+    are marked differently, the refinement flag takes priority.
+
+    Both refinement and coarsening flags can only be set of leaves.
+    To prevent drastic mesh change, coarsening is only executed when a leaf
+    box is marked for coarsening, and its parent's children are all leaf
+    boxes (so that change in the number of boxes is bounded per box flagged).
+    Please note that the above behavior may be subject to change in the future.
+
+    :arg refine_flags: a boolean array of size `nboxes`.
+    :arg coarsen_flags: a boolean array of size `nboxes`.
+    :arg error_on_ignored_flags: if true, an exception is raised when enforcing
+        level restriction requires ignoring some coarsening flags.
+    :returns: a processed copy of the tree.
+    """
+    if refine_flags is None:
+        refine_flags = np.zeros(tob.nboxes, bool)
+    if coarsen_flags is None:
+        coarsen_flags = np.zeros(tob.nboxes, bool)
+
+    if (refine_flags & coarsen_flags).any():
+        raise ValueError("some boxes are simultaneously marked "
+                         "to refine and coarsen")
+
+    tob = _apply_refine_flags_without_sorting(refine_flags, tob)
+    coarsen_flags = _resized_array(coarsen_flags, tob.nboxes)
+    tob = _apply_coarsen_flags(coarsen_flags, tob, error_on_ignored_flags)
+    return _sort_and_prune_deleted_boxes(tob)
+
+# }}} End refine/coarsen a tree of boxes
+
+
+def make_tob_root(dim: int,
+                  bbox: Union[npt.NDArray, List[List[float]]]) -> TreeOfBoxes:
+    """
+    Make the minimal tree of boxes.
+    """
+    center = np.array(
+        [(bbox[0][iaxis] + bbox[1][iaxis]) * 0.5 for iaxis in range(dim)])
+    extents = np.array(
+        [(bbox[1][iaxis] - bbox[0][iaxis]) for iaxis in range(dim)])
+    from pytools import single_valued
+    parents = np.array([0], np.intp)
+    parents[0] = -1  # root has no parent
+    return TreeOfBoxes(
+            box_centers=center.reshape(dim, 1),
+            root_extent=single_valued(extents, np.allclose),
+            box_parent_ids=parents,
+            box_child_ids=np.array([0] * 2**dim, np.intp).reshape(2**dim, 1),
+            box_levels=np.array([0]),
+            )
+
+
+def make_mesh_from_leaves(tob: TreeOfBoxes) -> "Mesh":
```

Comment:
Should maybe also return a vector providing the element number $\to$ box number mapping (or inverse of that, with holes).